migrations/migration.go

```diff
@@ -303,34 +318,35 @@ func (m *StorageMigration) MigrateNestedValue(
 			if newKey == nil {
 				keyToSet = existingKey
 			} else {
-				// Key was migrated.
-				// Remove the old value at the old key.
-				// This old value will be inserted again with the new key, unless the value is also migrated.
-				existingKey = legacyKey(existingKey)
-				oldValue := dictionary.Remove(
-					m.interpreter,
-					emptyLocationRange,
-					existingKey,
-				)
+				keyToSet = newKey
+			}
 
-				if _, ok := oldValue.(*interpreter.SomeValue); !ok {
-					panic(errors.NewUnexpectedError(
-						"failed to remove old value for migrated key: %s",
-						existingKey,
-					))
-				}
+			existingKey = legacyKey(existingKey)
+			existingKeyStorable, existingValueStorable := dictionary.RemoveWithoutTransfer(
+				m.interpreter,
+				emptyLocationRange,
+				existingKey,
+			)
 
-				keyToSet = newKey
+			if existingKeyStorable == nil {
+				panic(errors.NewUnexpectedError(
+					"failed to remove old value for migrated key: %s",
+					existingKey,
+				))
 			}
 
 			if newValue == nil {
 				valueToSet = existingValue
 			} else {
 				// Value was migrated
 				valueToSet = newValue
+
+				interpreter.StoredValue(m.interpreter, existingValueStorable, m.storage).
+					DeepRemove(m.interpreter)
+				m.interpreter.RemoveReferencedSlab(existingValueStorable)
```

Comment:
Shouldn't this be done despite the 'value' was migrated or not? i.e: key could be migrated and the old value may now needs to be inserted at a different key?

Reply:
We should only delete the data for the value, if there is a new value. If we are re-inserting the value for a different key, then removing the data for the value would loose the data.

Note also the change above: We're now _always_ removing the value from the container (but that does not imply deletion), i.e. after the old key-value pair is no longer in the container, we can then either:
- Insert the existing value under the new key. We shouldn't delete the value 
- Insert the new value under the (old or new) key. We need to delete the old value

You can try making the change, and I tried so myself yesterday: If the old value is always deleted, then atree will report the missing data (slab).

Reply:
Thanks for the explanation!